File: justdb-parser/src/main/java/com/lfkdsk/justdb/parser/SQLParser.java
package com.lfkdsk.justdb.parser;/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import bnfgenast.ast.base.AstList;
import bnfgenast.ast.base.AstNode;
import bnfgenast.bnf.BnfCom;
import com.lfkdsk.justdb.parser.lexer.SQLLexer;
import com.lfkdsk.justdb.parser.literal.IDLiteral;
import com.lfkdsk.justdb.parser.literal.NumberLiteral;
import com.lfkdsk.justdb.parser.literal.StringLiteral;
import com.lfkdsk.justdb.parser.model.*;

import java.io.LineNumberReader;
import java.io.StringReader;

import static bnfgenast.ast.token.Token.EOL;
import static bnfgenast.bnf.BnfCom.leaf;
import static bnfgenast.bnf.BnfCom.rule;

public class SQLParser {
    private static final SQLLexer lexer = new SQLLexer("");
    public static final BnfCom<NumberLiteral> number = leaf(NumberLiteral::new).number();
    public static final BnfCom<StringLiteral> string = leaf(StringLiteral::new).string();
    public static final BnfCom<IDLiteral> id = leaf(IDLiteral::new).identifier(lexer.getReservedToken());
    public static final BnfCom<DataType> enumType = rule(DataType::new).or(
            rule().token("int"),
            rule().token("varchar").sep("(").ast(number).sep(")")
    );
    public static final BnfCom<Constant> constant = rule(Constant::new).or(number, string);
    public static final BnfCom<ConstantList> constantList = rule(ConstantList::new)
            .ast(constant)
            .repeat(rule().sep(",").ast(constant));
    public static final BnfCom<IDLiteral> field = id;
    public static final BnfCom<Expr> expr = rule(Expr::new).or(field, constant);
    public static final BnfCom<Term> term = rule(Term::new).ast(expr).sep("=").ast(expr);
    public static final BnfCom<FieldList> fieldList = rule(FieldList::new).ast(id).repeat(rule().sep(",").ast(id));
    public static final BnfCom<FieldDef> fieldDef = rule(FieldDef::new).ast(id).then(enumType);
    public static final BnfCom<FieldDefList> fieldDefList = rule(FieldDefList::new).ast(fieldDef)
            .repeat(rule().sep(",").ast(fieldDef));
    public static final BnfCom<CreateTable> createTable = rule(CreateTable::new)
            .sep("create").sep("table")
            .ast(id)
            .sep("(").then(fieldDefList).sep(")");
    public static final BnfCom<Insert> insert = rule(Insert::new)
            .sep("insert").sep("into")
            .ast(id)
            .sep("(").ast(fieldList).sep(")")
            .sep("values")
            .sep("(").ast(constantList).sep(")");

    public static final BnfCom<Predicate> predicate = rule(Predicate::new)
            .ast(term)
            .repeat(rule().sep("and").ast(term));

    public static final BnfCom<TableNameList> tableNameList = rule(TableNameList::new)
            .ast(id)
            .repeat(rule().sep(",").ast(id));

    public static final BnfCom<SelectList> selectList = rule(SelectList::new)
            .ast(field)
            .repeat(rule().sep(",").ast(field));

    public static final BnfCom<Query> query = rule(Query::new)
            .sep("select")
            .ast(selectList)
            .sep("from")
            .ast(tableNameList)
            .maybe(rule().sep("where").ast(predicate));

    public static final BnfCom<CreateView> createView = rule(CreateView::new)
            .sep("create").sep("view")
            .ast(id)
            .sep("as").then(query);

    public static final BnfCom<CreateIndex> createIndex = rule(CreateIndex::new)
            .sep("create").sep("index")
            .ast(id)
            .sep("on")
            .ast(id)
            .sep("(").ast(field).sep(")");

    public static final BnfCom<Delete> delete = rule(Delete::new)
            .sep("delete").sep("from")
            .ast(id)
            .maybe(rule().sep("where").ast(predicate));

    public static final BnfCom<Update> update = rule(Update::new)
            .sep("update")
            .ast(id)
            .sep("set")
            .ast(field).sep("=").ast(expr)
            .maybe(rule().sep("where").ast(predicate));

    public static final BnfCom<Create> create = rule(Create::new)
            .or(createTable, createView, createIndex).sep(";");

    public static final BnfCom<Modify> modify = rule(Modify::new)
            .or(insert, delete, update, query).sep(";");

    public static final BnfCom<AstList> program = rule(AstList::new).or(create, modify).sep(EOL);

    public static AstNode parse(String input) {
        lexer.setReader(new LineNumberReader(new StringReader(input)));
        return program.parse(lexer.tokens());
    }
}

File: src/sql/parser/SqlParser.kt
package sql.parser

import sql.ast.*
import sql.literal.EnumLeaf
import sql.literal.IdLeaf
import sql.literal.NumLeaf
import sql.literal.StrLeaf
import sql.token.SqlToken
import utils.parsertools.ast.AstNode
import utils.parsertools.combinators.Bnf
import utils.parsertools.exception.ParseException
import utils.parsertools.lex.Lexer


/**
 * Created by liufengkai on 2017/4/25.
 */

class SqlParser {
	// word set
	var reserved: HashSet<String> = HashSet()

	var typeDefSet: HashSet<String> = HashSet()

	val number = Bnf.rule().number(NumLeaf::class.java)

	val string = Bnf.rule().string(StrLeaf::class.java)

	var id = Bnf.rule().identifier(reserved, IdLeaf::class.java)

	val enum = Bnf.rule().enum(typeDefSet, EnumLeaf::class.java)

	val constant = Bnf.rule().or(number, string)

	val constantList = Bnf.rule(ConstantList::class.java)
			.ast(constant)
			.repeat(Bnf.rule().sep(",").ast(constant))

	val field = id

	val expr = Bnf.rule().or(field, constant)

	val term = Bnf.rule(Term::class.java)
			.ast(expr).sep("=").ast(expr)

	val fieldList = Bnf.rule()
			.identifier(reserved)
			.repeat(Bnf.rule().sep(",").identifier(reserved))

//	val typeDef = Bnf.rule().enum(reserved)

	val fieldDef = Bnf.rule(FieldList::class.java).identifier(reserved)
			.ast(enum)

	val fieldDefList = Bnf.rule(FieldDefList::class.java)
			.ast(fieldDef)
			.repeat(Bnf.rule().sep(",").ast(fieldDef))

	val createTable = Bnf.rule(CreateTable::class.java)
			.sep("create")
			.sep("table")
			.identifier(reserved)
			.sep("(")
			.ast(fieldDefList)
			.sep(")")

	val insert = Bnf.rule(Insert::class.java)
			.sep("insert", "into")
			.identifier(reserved)
			.sep("(")
			.ast(fieldList)
			.sep(")")
			.sep("values")
			.sep("(")
			.ast(constantList)
			.sep(")")

	val predicate = Bnf.rule(Predicate::class.java)
			.ast(term)
			.repeat(Bnf.rule().sep("and").ast(term))

	val tableNameList = Bnf.rule(TableNameList::class.java)
			.identifier(reserved)
			.repeat(Bnf.rule().sep(",").identifier(reserved))

	val selectList = Bnf.rule(SelectList::class.java)
			.ast(field)
			.repeat(Bnf.rule().sep(",").ast(field))

	val query = Bnf.rule(Query::class.java)
			.sep("select")
			.ast(selectList)
			.sep("from")
			.ast(tableNameList)
			.maybe(Bnf.rule().sep("where").ast(predicate))

	val createView = Bnf.rule(CreateView::class.java)
			.sep("create", "view")
			.identifier(reserved)
			.sep("as")
			.ast(query)

	val createIndex = Bnf.rule(CreateIndex::class.java)
			.sep("create", "index")
			.identifier(reserved)
			.sep("on")
			.identifier(reserved)
			.sep("(")
			.ast(field) // TODO fields?
			.sep(")")

	val delete = Bnf.rule(Delete::class.java)
			.sep("delete", "from")
			.identifier(reserved)
			.maybe(Bnf.rule().sep("where").ast(predicate))

	val update = Bnf.rule(Update::class.java)
			.sep("update")
			.identifier(reserved)
			.sep("set")
			.ast(field)
			.sep("=")
			.ast(expr)
			.maybe(Bnf.rule().sep("where").ast(predicate))

	val create = Bnf.rule().or(
			createTable.sep(";"),
			createIndex.sep(";"),
			createView.sep(";")
	)

	val modify = Bnf.rule().or(
			insert.sep(";"),
			delete.sep(";"),
			update.sep(";"),
			query.sep(";")
	)

	val singleCommand = Bnf.rule().or(create, modify)

	val program = Bnf.rule().ast(singleCommand).sep(SqlToken.EOL)

	init {
		// init type def params
		typeDefSet.add("VARCHAR")
		typeDefSet.add("INT")
	}

	@Throws(ParseException::class)
	fun parse(lexer: Lexer): AstNode {
		return program.parse(lexer)
	}


}